Add recursive argument to spack develop

psakievich opened the pull request:
This effort allows for a recursive develop call
which will traverse from the develop spec given back to the root(s) and mark all packages along the path as develop.

If people are doing development across the graph then paying fetch and full rebuild costs every time `spack install` is called is unnecessary and expensive. Since this requires an already concrete environment the exact versions can be used and there is no need to guess or provide versions for concrete specs.

Additional changes:

- Remove the constraint of only one develop spec.  Users can specify more than one unless they pass spec specific configurations like `--path` or `--build-directory`.
- Change the default check for the version to make a non-concrete CLI spec concrete.  The order now is:
   1. If the environment is concrete take the version in the already concretized environment (highest version, if there happens to be repeats)
   2. Next take `max(package_class.versions)`


<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

psakievich committed:
Add recursive argument to spack develop

This effort allows for a recursive develop call
which will traverse from the develop spec given back to the root(s)
and mark all packages along the path as develop.

If people are doing development across the graph then paying
fetch and full rebuild costs every time spack develop is called
is unnecessary and expensive.

Also remove the constraint for concrete specs and simply take the
max(version) if a version is not given. This should default to the
highest infinity version which is also the logical best guess for
doing development.

tgamblin commented:
Can you split this into two PRs -- one for the removal of the constraint and the other for the recursive option?

psakievich commented:
I put them together because the recursive makes sense to default to just the package name since it is already concretized. So we should do remove constraint first and then recursive dependent on that merging IMO.

psakievich commented:
@tgamblin see #46911. I preemptively added you as a reviewer 🙂

psakievich committed:
Merge branch 'develop' into psakiev/recursive-develop

psakievich committed:
Completion commands

psakievich committed:
Add docs

psakievich commented:
I started a unit-test for this, but it's looking like a big lift at the moment. It needs a concretized environment to work, and so far all I've found is mock concretization of a spec. Seems like a notable gap in our testing infrastructure but would dramatically extend the scope of this PR to implement and should be a PR in and of itself IMHO.

psakievich committed:
Updates after testing

psakievich commented:
Testing this locally for now:

``` console
$ spack env activate --temp 
$ spack add exawind
$ spack concretize
$ spack develop -r trilinos
# trilinos actually has a fetch error that I've documented but never fixed #25370 
$ spack develop -r trilinos
# clones see yaml file below
```

``` yaml
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  specs:
  - exawind
  view: true
  concretizer:
    unify: true
  develop:
    nalu-wind:
      spec: nalu-wind@=master
    exawind:
      spec: exawind@=master
    trilinos:
      spec: trilinos@=16.0.0
```

psakievich commented:
@haampie this should meet your `spack develop` needs as outlined on slack.

haampie commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -123,6 +131,25 @@ def develop(parser, args):
+                "Recursive develop requires a concretized environment"
+            )
+        else:
+            for s in concrete_specs:
+                for parent in s.traverse_edges(direction="parents", root=True):
```
Use spack.traverse.traverse_nodes(concrete_specs) instead of a double loop and key by dag hash to uniquely visit each node. Your current code visits nodes repeatedly.

I don't think it applies here, but you have to be very careful with upwards traversal, since sometimes it follows edges from parents that are no descendants of any of concrete_specs. That happens when specs are taken from the db (where all in-edges of the entire db are added).

psakievich replied:
Do you think there is a better traversal logic I could get to hit the intended intersection of the DAG? I read the warnings about upward traversal, and I do find it concerning.  I could see a visit each node in the DAG and check a condition if there is an accessible and sufficient condition on the specs to say "I depened on spec foo in my subgraph?".


Maybe something like:
``` python
if f"^{dev_spec.name}" in parent_spec.fullname:
   develop(parent_spec)
```

psakievich replied:
I think I'll mark this as resolved for now. I don't really know how to test this. I haven't hit an issue with it, but if someone wants to come up with some potential tests then I will be happy to run them.

haampie commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -123,6 +131,25 @@ def develop(parser, args):
+                    parent_args.recursive = False
+
+                    tty.debug(f"Recursive develop for {parent_args.spec}")
+                    develop(parser, parent_args)
```
Make the develop logic a function of its own that does not depend on argparse.

haampie commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -123,6 +131,25 @@ def develop(parser, args):
+                for parent in s.traverse_edges(direction="parents", root=True):
+
+                    parent_args = copy(args)
+                    parent_args.spec = parent.spec.format("{name}@{version}")
```
You want to use dag hashes... name@version is not specific enough.

psakievich replied:
We can't use dag hashes b/c they will change when we concretize. I went generic to allow hitting multiple instances. I have always tried to go most generic for dev specs out of principle. I could do the `spec.short_spec` but if someone has a development environment with repeat specs (+cuda/~cuda) then `spack develop` can't support both.

psakievich replied:
@haampie will you weigh back in on this? I'd like to get this merged.

haampie commented:
> @haampie this should meet your `spack develop` needs as outlined on slack.

Not really, I said I wanna be able to select an individual concrete spec from a concrete env:

<img width="1438" alt="Screenshot 2024-10-19 at 10 31 01" src="https://github.com/user-attachments/assets/99a97c0e-251c-4090-8c3a-066c7591e1b6">

^ should modify the concrete environment.

psakievich commented:
> So, in short: I wanna be able to add dev_path=... post-hoc, to a single spec selected from a concrete environment, without re-concretization.

Ah well I can add the single spec logic to this PR (prefer concrete if a version is omitted), but me thinks skipping re-concretization is a more fundamental change for another PR.

psakievich commented:
Thinking about this more I think to skip concretization we'd likely need to change how develop is recognized (make it an environment property rather than a spec property). It messes with a lot of fundamental stuff though. The other option I can think of is a surgical change to the DAG. Modify specs, change hashes, but don't actually solve.

haampie commented:
Note that `spack spec cmake ^curl dev_path=/tmp` works ;) which is already more convenient than the `develop: ...` section in an env. With that in mind the whole `develop: ...` section is kind of redundant to start with...

psakievich commented:
@haampie That order makes sense to me. The only reason it needs to be part of the spec with a variant is to mark for the install/future reuse. From and install point dev spec doesn't really matter? I guess it does in that you probably don't want to reuse a develop spec or rather don't want it masquerading as something it is not. Other than that I think it should be an environment property and not a spec property with the order you outlined. 

I do like that develop can happen pre concretization for reuse of developer environments etc, but I don't see why that would have to change. 

I suspect that the install situation could be alleviated by not installing develop specs or their dependents into the same install tree. Have it be a special one. That actually aligns well with another request of having develop specs link and possibly rebuild out of a build tree and skip installation all together. Rational being linking to build directly means tighter changes (update a header and not have to reinstall just rebuild) and less file system work.

psakievich committed:
Refactor develop

psakievich committed:
Refactor further

psakievich commented:
I made a few more updates.  I refactored so that the core of developing a spec is 3 steps:

1) confirm it is concrete and pick a default if it is not (default to highest version in environment if concrete env, otherwise max version)
2) setup the source code
3) update the config/spack.yaml

These steps should be taken regardless of if a spec is from recursive collection, reading and updating the current develop specs or just CLI specified specs.

So those differentiating cases were moved to a generator function. I also removed the constraint of only being able to list one spec as develop at the CLI, but preserve that if `--path` or `--build-directory` are passed.  Someday I would like those to be projections so then they could be spec aware/valid for the multi spec case.

psakievich commented:
I also can't seem to figure out why this unit-test keeps failing. I have tested updating a develop spec from the CLI and it works properly.

psakievich committed:
Fix unit-test

psakievich committed:
Merge branch 'develop' into psakiev/recursive-develop

psakievich committed:
Completion

psakievich committed:
Add unit test

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
+                if args.recursive:
+                    concrete_specs = env.all_matching_specs(spec)
+                    if not concrete_specs:
+                        tty.msg(
+                            "No matching specs found in the environment. "
+                            "Recursive develop requires a concretized environment"
```
This needs to say which spec didn't match, and should make clear that this spec is being skipped but others will still be developed.

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
-    else:
-        abspath = path
+    entry = {"spec": str(spec)}
+    if path != spec.name:
```
```suggestion
    if path and path != spec.name:
```

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
             )
-        _update_config(spec, path)
+        # add develop spec and update path
+        if specified_path:
+            # always overwrite path if it was supplied
+            _update_config(spec, specified_path)
+        else:
+            _update_config(spec, spec.name)
```
If the path is just defaulting to the spec name, we don't need to pass it in at all
```suggestion
        # always overwrite path if it was supplied
        _update_config(spec, specified_path)
```

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
+    """
+    tty.debug(f"Updating develop config for {env.name} transactionally")
+
+    # this check has to happen outside the write transaction otherwise updates don't seem to happen
```
Can you fill me in a bit on what was happening with this? I don't understand why it would be necessary and that concerns me.

psakievich replied:
This was just for unit-tests to pass. It worked from the CLI.  I think it had to do with process forks?  I'm not sure exactly. I spend a lot of time trying to debug but all my debugging instances were getting hung up in the subprocesses and the inspection/print statements looked like the values were correct, but the test always failed for updating the develop spec.

becker33 replied:
Which test was the one failing? I'm hesitant to just ignore this since it suggests something wrong with how we handle environments more broadly.

psakievich replied:
`test_develop_update_spec`

psakievich replied:
With the current PR if I move this section back inside the `write_transaction` the following tests fail:

``` console
======================================================= short test summary info =======================================================
FAILED lib/spack/spack/test/cmd/develop.py::TestDevelop::test_develop_update_spec[mock_archive0] - AssertionError: assert 'mpich@=1.0' == 'mpich@=2.0'
FAILED lib/spack/spack/test/cmd/develop.py::test_recursive[mock_archive0] - AssertionError: assert 'indirect-mpich' in {'direct-mpich': {'path': 'direct-mpich', 'spec': 'direct-mpich@1.0/ldxba4og2st3nqkukqp...
```

psakievich replied:
They pass if I check for the dev_entry outside of the `write_transaction`.

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
-                raise SpackError(msg)
 
-        _retrieve_develop_source(spec, abspath)
+def update_env_file(
```
I think this would be easier to read as `update_env` since it's interacting with the `Environment` through the normal `spack.config` interface, not the `EnvironmentManifestFile` interface

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
+        # first check environment for a matching concrete spec
+        matching_specs = env.all_matching_specs(spec)
+        if matching_specs:
+            version = matching_specs[0].version
```
This seems unsafe -- won't it cause problems for any matching spec other than the highest version?

psakievich replied:
Yes, but that is the consequence of only allowing one develop spec per package.  Perhaps we should just throw in this instance?

becker33 replied:
Yeah I think if we won't properly handle multiple specs with different versions, we should raise if we hit that.

psakievich replied:
That's the new behavior and I added a test for it.

psakievich committed:
Respond to review and add another test

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -84,86 +79,152 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
+                    if not concrete_specs:
+                        tty.warn(
+                            f"{spec.name} has no matching concrete specs in the environment and "
+                            "will be skipped. Recursive develop requires a concretized"
```
```suggestion
                            "will be skipped. `spack develop --recursive` requires a concretized"
```

psakievich commented:
@spackbot fix style

psakievich committed:
[@spackbot] updating style on behalf of psakievich

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -84,86 +79,152 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
+                if not test_spec.satisfies(m_spec):
+                    raise SpackError(
+                        f"{spec.name}: has conflicting specs and only one develop spec per package "
+                        "can be defined. Please unify the versions or reduce the specs to use "
+                        "`spack develop`"
```
This last sentence doesn't scan easily for me -- is the second suggestion supposed to be "or do not `spack develop` {spec.name}"?

psakievich replied:
Well no, but that is fine too. It's supposed to read, make the versions the same or delete specs until you only have the same version to use spack develop. I suppose that second one is redundant.

psakievich replied:
@becker33  I've updated the error message:
``` console
$ spack develop zlib-ng
==> Error: zlib-ng: has multiple concrete instances in the graph that can't be satisified by a single develop spec. To use `spack develop` ensure one of the following:
 a) zlib-ng nodes can satisfy the same develop spec (minimally this means they all share the same version)
 b) Provide a concrete develop spec (zlib-ng@[version]) to clearly indicate what should be developed
```

psakievich committed:
Update develop.py

Co-authored-by: Greg Becker <becker33@llnl.gov>

psakievich committed:
Improve error message

psakievich committed:
Merge branch 'develop' into psakiev/recursive-develop

psakievich commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -83,86 +78,156 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
-                msg += " because its path already exists."
-                tty.msg(msg)
-                continue
+def assure_concrete_spec(env: spack.environment.Environment, spec: spack.spec.Spec):
```
```suggestion
def assure_concrete_spec(env: "spack.environment.Environment", spec: spack.spec.Spec):
```

psakievich committed:
Str for type checking

psakievich committed:
More str type check conversions

psakievich committed:
Use the import over string for type check

becker33 committed:
env.dev_specs: remove caching for computed dev_specs

this cached value is not properly invalidated when develop config is
written, as in the `spack develop` command. Some accesses are accidentally
correct because a read or write transaction on the environment will
invalidate the `self._dev_specs` as a side-effect. Considered adding
invalidation but determined it was not worth the effort to bridge the gap in
interface between `spack.config` and `spack.environment`.

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -38,28 +40,22 @@ def setup_parser(subparser):
         action="store_true",
         dest="clone",
-        default=None,
+        default=True,
```
I don't understand this change -- this seems to suggest in the default case we clone regardless of whether the path is already cloned?

psakievich replied:
This is suggesting we call the `_clone` method and attempt to clone. Additional checks are there to explain why we skip cloning.  It's been a while but IIRC the logic was embedded off state of `None` `True` or `False` before.  But for a recursive clone we need better inferences that I moved inside a method.

psakievich replied:
Need to change the help messages.

psakievich replied:
@becker33 updated this one

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -83,86 +79,154 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
+    Generator function to loop over all the develop specs based on how the command is called
+    If no specs are supplied then loop over the develop specs listed in the environment.
+    """
+    if not args.spec:
+        if args.clone is False:
+            raise SpackError("No spec provided to spack develop command")
+        else:
+            for name, entry in env.dev_specs.items():
+                path = entry.get("path", name)
+                abspath = spack.util.path.canonicalize_path(path, default_wd=env.path)
+                # Both old syntax `spack develop pkg@x` and new syntax `spack develop pkg@=x`
+                # are currently supported.
+                spec = spack.spec.parse_with_version_concrete(entry["spec"])
+                yield spec, abspath
+    else:
```
```suggestion
    if not args.spec:
        if args.clone is False:
            raise SpackError("No spec provided to spack develop command")
        for name, entry in env.dev_specs.items():
            path = entry.get("path", name)
            abspath = spack.util.path.canonicalize_path(path, default_wd=env.path)
            # Both old syntax `spack develop pkg@x` and new syntax `spack develop pkg@=x`
            # are currently supported.
            spec = spack.spec.parse_with_version_concrete(entry["spec"])
            yield spec, abspath
        return
```
This is purely cosmetic, to reduce the amount of indentation needed and make the code more legible. With this change then the entire rest of the function can be dedented by 4 spaces.

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -83,86 +79,154 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
+                "spack develop requires at most one named spec when using the --path or"
+                " --build-directory arguments"
+            )
+        else:
```
Similarly, can we remove this `else` statement since the `if` block exits and dedent the entire else case?

psakievich committed:
Cosmetic updates

psakievich commented:
@spackbot fix style

psakievich committed:
[@spackbot] updating style on behalf of psakievich

becker33 commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -83,86 +79,154 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
+            version = matching_specs[0].version
+            test_spec = spack.spec.Spec(f"{spec}@{version}")
+            for m_spec in matching_specs:
+                if not test_spec.satisfies(m_spec):
```
I think this needs to be
```suggestion
                if not test_spec.intersects(m_spec):
```

psakievich replied:
The order was wrong here. we should be checking that all the matching specs in the environment satisfy the test spec which was constructed from the name and version of the first matching spec.

psakievich committed:
Fix logic misses

psakievich committed:
Merge branch 'develop' into psakiev/recursive-develop

psakievich commented:
@spackbot fix style

psakievich committed:
[@spackbot] updating style on behalf of psakievich

psakievich committed:
Update help messages

psakievich committed:
Indicate default behavior

psakievich commented:
@spackbot fix style

psakievich committed:
style

psakievich committed:
wild stylz

psakievich commented:
@spackbot fix style

psakievich committed:
[@spackbot] updating style on behalf of psakievich

psakievich committed:
Completion
